### ISSUE TITLE
TestCase.assertEqual does not respect longMessage

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#76813]
### 🐛 Describe the bug

In stock unittest, setting longMessage = True on a class means that explicitly specifying a msg on an assertEqual call should also still preserve the automatically generated message. This does not happen with PyTorch's assertEqual

```
self.longMessage = True
self.assertEqual(torch.int64, torch.int32, msg="boo")
```

### Versions

master

**Comment by ezyang** [comment]
cc @pmeier @mruberry

**Comment by pmeier** [comment]
For reference: https://docs.python.org/3/library/unittest.html#unittest.TestCase.longMessage

**Comment by pmeier** [comment]
`git grep`ing through the repository revealed two usages

https://github.com/pytorch/pytorch/blob/563c2719bff5a13d7cd1e3c492a4d01edfc307eb/test/test_decomp.py#L336-L337

https://github.com/pytorch/pytorch/blob/563c2719bff5a13d7cd1e3c492a4d01edfc307eb/test/distributed/elastic/rendezvous/utils_test.py#L360

We only removed the ability for custom "runtime" messages in #67794. Previously, we accepted a callable for the `msg` that got passed diagnostics to compose the message manually. Since #67794 opened up the comparison logic for more types than just tensors, this was no longer possible.

Still, we could reinstate the callable `msg` argument which will get passed our internal message and return a new one. With that the `TestCase.longMessage` can be achieved by passing

```py
msg = lambda orig_msg: f"{orig_msg}\n\n{msg}" if msg and self.longMessage else msg
```

from `TestCase.assertEqual`.
